src/coreclr/jit/importer.cpp

```diff
@@ -4167,20 +4169,59 @@ GenTree* Compiler::impImportStaticReadOnlyField(CORINFO_FIELD_HANDLE field, CORI
             GenTree* cnsValue = impImportCnsTreeFromBuffer(buffer, fieldType);
             if (cnsValue != nullptr)
             {
+                JITDUMP("... success! The value is:\n");
+                DISPTREE(cnsValue);
                 return cnsValue;
             }
         }
     }
     else if (fieldType == TYP_STRUCT)
     {
+        unsigned totalSize = info.compCompHnd->getClassSize(fieldClsHnd);
+        unsigned fieldsCnt = info.compCompHnd->getClassNumInstanceFields(fieldClsHnd);
 
-        if (info.compCompHnd->getClassNumInstanceFields(fieldClsHnd) != 1)
+        // For large structs we only want to handle "initialized with zero" case
+        // e.g. Guid.Empty and decimal.Zero static readonly fields.
+        if ((totalSize > TARGET_POINTER_SIZE) || (fieldsCnt != 1))
         {
-            // Only single-field structs are supported here to avoid potential regressions where
-            // Metadata-driven struct promotion leads to regressions.
+            JITDUMP("checking if we can do anything for a large struct ...");
+            const int MaxStructSize = 64;
+            if ((totalSize == 0) || (totalSize > MaxStructSize))
+            {
+                // Limit to 64 bytes for better throughput
+                JITDUMP("struct is larger than 64 bytes - bail out.");
+                return nullptr;
+            }
+
+            uint8_t buffer[MaxStructSize] = {0};
+            if (info.compCompHnd->getReadonlyStaticFieldValue(field, buffer, totalSize))
+            {
+                for (unsigned i = 0; i < totalSize; i++)
+                {
+                    if (buffer[i] != 0)
+                    {
+                        // Value is not all zeroes - bail out.
+                        // Although, We might eventually support that too.
+                        JITDUMP("value is not all zeros - bail out.");
+                        return nullptr;
+                    }
+                }
+
+                JITDUMP("success! Optimizing to ASG(struct, 0).");
+                unsigned structTempNum = lvaGrabTemp(true DEBUGARG("folding static ro fld empty struct"));
+                lvaSetStruct(structTempNum, fieldClsHnd, false);
+                GenTreeLclVar* tree = gtNewLclvNode(structTempNum, TYP_STRUCT);
+                impAppendTree(gtNewBlkOpNode(tree, gtNewIconNode(0), false, false), CHECK_SPILL_NONE, impCurStmtDI);
+                return gtNewLclvNode(structTempNum, TYP_STRUCT);
```

Comment:
We can get here for vectors, presumably, plain `TYP_STRUCT` would not be right for them.

Reply:
@SingleAccretion this whole block is guarded with `else if (fieldType == TYP_STRUCT)` so presumably it should not happen?

Reply:
But the field type is not normalized as far as I see. `getFieldType` will return `TYP_STRUCT` for vectors.

Reply:
Ah, you're right and it actually asserts, fixed 🙂